packages/plugin-workspaces-tools/sources/commands/foreach.ts

```diff
@@ -0,0 +1,187 @@
+import { Configuration, LocatorHash, PluginConfiguration, Project, Workspace } from '@berry/core';
+import { structUtils, Cache, DescriptorHash, StreamReport }                    from '@berry/core';
+import { PassThrough, Writable }                                               from 'stream';
+import { cpus }                                                                from 'os';
+import chalk                                                                   from 'chalk';
+// @ts-ignore
+import pLimit                                                                  from 'p-limit';
+
+
+type ForeachOptions = {
+  args: Array<string>;
+  command: string;
+  cwd: string;
+  exclude: string[];
+  include: string[];
+  interlaced: boolean;
+  parallel: boolean;
+  prefixed: boolean;
+  stdout: Writable;
+  withDependencies: boolean;
+}
+
+// eslint-disable-next-line arca/no-default-export
+export default (clipanion: any, pluginConfiguration: PluginConfiguration) => clipanion
+
+    .command(`workspaces foreach <command> [...args] [-p,--parallel] [--with-dependencies] [-I,--interlaced] [-P,--prefixed] [-i,--include WORKSPACES...] [-x,--exclude WORKSPACES...]`)
+    .flags({proxyArguments: true})
+
+    .categorize(`Workspace-related commands`)
+    .describe(`run a command on all workspaces`)
+
+    .action(
+      async ({cwd, args, stdout, command, exclude, include, interlaced, parallel, withDependencies, prefixed, ...env}: ForeachOptions) => {
+        const configuration = await Configuration.find(cwd, pluginConfiguration);
+        const { project } = await Project.find(configuration, cwd);
+        const cache = await Cache.find(configuration);
+
+        const needsProcessing = new Map<LocatorHash, Workspace>();
+        const processing = new Set<DescriptorHash>();
+        const concurrency = parallel ? Math.max(1, cpus().length / 2) : 1;
+        const limit = pLimit(concurrency);
+        let commandCount = 0;
+
+        const report = await StreamReport.start({configuration, stdout}, async report => {
+          await project.resolveEverything({lockfileOnly: true, cache, report});
+
+          let workspaces = command.toLowerCase() === `run`
+            ? project.workspaces.filter(workspace => workspace.manifest.scripts.has(args[0]))
+            : project.workspaces;
+
+          if (include.length > 0) {
+            workspaces = workspaces.filter(workspace => include.includes(workspace.locator.name))
+          }
+
+          if (exclude.length > 0) {
+            workspaces = workspaces.filter(workspace => !exclude.includes(workspace.locator.name))
+          }
+
+          for (const workspace of workspaces) {
+            needsProcessing.set(workspace.locator.locatorHash, workspace);
+          }
+
+          while (needsProcessing.size > 0) {
```

Comment:
This function basically runs [Kahn's algorithm](https://en.wikipedia.org/wiki/Topological_sorting#Kahn's_algorithm), which results in suboptimal timings for `--parallel --with-dependencies`. Lerna also suffers from this.

An example as proof:

Suppose I've got seven packages: `a` through `f` and `x`. Packages `a` through `f` don't depend on anything, `x` depends on `a`.

We'll loop twice through this loop.
The first iteration will run packages `a`, `b`, `c`, `d`, `e` and `f`.
The second iteration will run package `x`

Let's assume a quadcore with hyperthreading, we'll run four packages in parallel. For simplicity we'll say the script takes 10s in every package.
The timings with the current code would be:

- `T`: start packages `a`, `b`, `c` and `d`
- `T + 10s`: start packages `e` and `f`
- `T + 20s`: start package `x`
- `T + 30s`: finished

While the optimal timings on that same machine would be

```
- `T`: start packages `a`, `b`, `c` and `d`
- `T + 10s`: start packages `e`, `f` and `x`
- `T + 20s`: finished
```

---

I propose turning this around. Instead of looping through all packages and checking which are ready to pick up then waiting for those to finish and looping again, we store a mapping of dependencies to dependant and pick up the packages that don't have any open dependencies anymore

I've run an algorithm basically like this on a codebase at work and noticed significant improvements in the time it takes to run through all workspaces. I don't remember the timings exactly but I could run it tomorrow if you want to have them.

Reply:
That's actually something we discussed with @deini - his initial implementation started a new candidate process as soon as one job ended. I suggested to use the loop for now, as we're already using a similar algorithm when building the packages and it's a fairly easy algorithm to read and understand.

That said, if it improves the perfs significantly in real life conditions, it might make sense to later refactor both this and the build scripts to use an improved algorithm.

Reply:
> if it improves the perfs significantly in real life conditions

I've been running tests, I get a speedup of about 4% (15s on a total of 390-405s). Running this test on my workspace makes my system load peak at 25 for an 8 CPU thread machine though, so I'm guessing the speedup would be more significant for scripts that don't require so much power. I don't have any real life versions of those on hand though.

Knowing this, I agree that the easier algorithm outweighs the speedup benefit until proof of more significant slowdowns is provided